agent/pkg/singularity/singularity.go

```diff
@@ -0,0 +1,613 @@
+		if wErr := archive.Write(src, a.Archive, func(level, log string) error {
+			return p.Publish(ctx, docker.NewLogEvent(level, log))
+		}); wErr != nil {
+			return nil, fmt.Errorf("writing archive for %s: %w", a.Path, err)
+		}
+	}
+
+	// Do not mount top level dirs that are likely to conflict inside of the container, since
+	// these mounts do not overlay. Instead, mount their children.
+	ignoredPathPrefixes := []string{"/", "/etc", "/opt", "/run", "/etc/ssh"}
+	var mountPoints []string
+	// This depends on walkdir walking in lexical order, which is documented.
+	if wErr := filepath.WalkDir(tmpdir, func(src string, d fs.DirEntry, err error) error {
+		p := strings.TrimPrefix(src, tmpdir)
+
+		// If an existing mount point covers this path, nothing to add
+		for _, m := range mountPoints {
+			if strings.HasPrefix(p, m) {
+				return nil
+			}
+		}
+
+		dirPaths := filepath.SplitList(p)
+		prefix := ""
+		// Search to find the top-most unmounted path
+		for i := 0; i < len(dirPaths); i++ {
+			prefix = filepath.Join(prefix, dirPaths[i])
+
+			s.log.Trace("Checking mountPoint prefix {}", prefix)
+			if !slices.Contains(ignoredPathPrefixes, prefix) {
+				s.log.Trace("Add mountPoint {}", prefix)
+				mountPoints = append(mountPoints, prefix)
+				return nil
+			}
+		}
+		s.log.Warnf("could not determine where to mount %s", src)
+		return nil
+	}); wErr != nil {
+		return nil, fmt.Errorf("determining mount points: %w", err)
+	}
+	for _, m := range mountPoints {
+		args = append(args, "--bind", fmt.Sprintf("%s:%s", path.Join(tmpdir, m), m))
+	}
+
+	for _, m := range req.HostConfig.Mounts {
+		// TODO(singularity): Investigate handling these options.
+		if m.ReadOnly {
+			if pErr := p.Publish(ctx, docker.NewLogEvent(model.LogLevelWarning, fmt.Sprintf(
+				"mount %s:%s was requested as readonly but singularity does not support this; "+
+					"will bind mount anyway, without it being readonly",
+				m.Source, m.Target,
+			))); pErr != nil {
+				return nil, pErr
+			}
+		}
+		if m.BindOptions != nil && m.BindOptions.Propagation != "rprivate" { // rprivate is default.
+			if pErr := p.Publish(ctx, docker.NewLogEvent(model.LogLevelWarning, fmt.Sprintf(
+				"mount %s:%s had propagation settings but singularity does not support this; "+
+					"will bind mount anyway, without them",
+				m.Source, m.Target,
+			))); pErr != nil {
+				return nil, pErr
+			}
+		}
+		args = append(args, "--bind", fmt.Sprintf("%s:%s", m.Source, m.Target))
+	}
+
+	if shmsize := req.HostConfig.ShmSize; shmsize != 4294967296 { // 4294967296 is the default.
+		if pErr := p.Publish(ctx, docker.NewLogEvent(model.LogLevelWarning, fmt.Sprintf(
+			"shmsize was requested as %d but singularity does not support this; "+
+				"we do not launch with `--contain`, so we inherit the configuration of the host",
+			shmsize,
+		))); pErr != nil {
+			return nil, pErr
+		}
+	}
+
+	// TODO(singularity): Un-dockerize the RunContainer API so we can know to pass `--rocm` without
```

Comment:
i will ticket these in a follow up and try to do something cleaner. i think these things just need to get set at a lower level. it is all set in the master because the master assumes docker right now.